File: Cpp/1673_find_the_most_competitive_subsequence.cpp
// https://leetcode.cn/problems/find-the-most-competitive-subsequence/

class Solution {
public:
    vector<int> mostCompetitive(vector<int>& nums, int k) {
        vector<int> res;
        int n = nums.size();
        for (int i = 0; i < n; ++i) {
            int num = nums[i];
            while (res.size() > 0 && num < res.back() && res.size() + n - i > k) {
                res.pop_back();
            }
            res.push_back(num);
        }
        res.resize(k);
        return res;
    }
};

File: Python3/1673_find_the_most_competitive_subsequence.py
# https://leetcode.cn/problems/find-the-most-competitive-subsequence/

class Solution:
    def mostCompetitive(self, nums, k):
        """
        Inputs:
            nums: list[int]
            k: int
        Outputs:
            res: list[int]
        """
        res = []
        n = len(nums)
        for i, num in enumerate(nums):
            while len(res) > 0 and num < res[-1] and len(res) + n - i > k:
                res.pop()
            res.append(num)
        return res[:k]